crates/sui-graphql-rpc/src/types/dynamic_field.rs

```diff
@@ -0,0 +1,189 @@
+    base64::Base64, move_object::MoveObject, move_value::MoveValue, sui_address::SuiAddress,
+};
+use crate::context_data::package_cache::PackageCache;
+use crate::error::{code, Error};
+use crate::{context_data::db_data_provider::PgManager, error::graphql_error};
+use sui_types::object::Object as NativeSuiObject;
+
+pub(crate) struct DynamicField {
+    pub stored_object: StoredObject,
+}
+
+impl DynamicField {
+    pub fn is_dynamic_object(&self) -> bool {
+        self.stored_object.df_kind == Some(DynamicFieldType::DynamicObject as i16)
+    }
+
+    pub fn dynamic_field_type(&self) -> Result<DynamicFieldType> {
+        match self.stored_object.df_kind {
+            None => Err(Error::Internal("Dynamic field type is not set".to_string()).extend()),
+            Some(df_kind) => match df_kind {
+                0 => Ok(DynamicFieldType::DynamicField),
+                1 => Ok(DynamicFieldType::DynamicObject),
+                _ => Err(Error::Internal("Unexpected df_kind value".to_string()).extend()),
+            },
+        }
+    }
+}
+
+pub(crate) async fn deserialize_move_struct(
+    serialized_object: &[u8],
+    cache: &PackageCache,
+) -> Result<(StructTag, MoveStruct)> {    
+    let native_object: NativeSuiObject = bcs::from_bytes(serialized_object)
+        .map_err(|e| Error::Internal(format!("Failed to deserialize object: {e}")).extend())?;
+    let move_object = native_object.data.try_as_move().ok_or_else(|| {
+        Error::Internal("Failed to convert object into MoveObject".to_string()).extend()
+    })?;
+    let struct_tag = native_object
+        .data
+        .struct_tag()
+        .ok_or_else(|| Error::Internal("StructTag missing on object".to_string()).extend())?;    
+    let contents = move_object.contents();
+    let type_tag = TypeTag::Struct(Box::new(struct_tag.clone()));    
+    let move_type_layout = cache.type_layout(type_tag.clone()).await?;
+    let move_struct = match move_type_layout {
+        MoveTypeLayout::Struct(move_struct_layout) => {
+            MoveStruct::simple_deserialize(contents, &move_struct_layout)
+        }
+        _ => Err(Error::Internal("Object is not a move struct".to_string()).extend())?,
+    }?;
+    Ok((struct_tag, move_struct))
+}
+
+pub fn extract_field_from_move_struct(
+    move_struct: MoveStruct,
+    field_name: &str,
+) -> Result<value::MoveValue> {
+    match move_struct {
+        MoveStruct::WithTypes { fields, .. } => {
+            fields.into_iter().find_map(|(id, value)| {
+                if id.to_string() == field_name {
+                    Some(value)
+                } else {
+                    None
+                }
+            })
+        }
+        .ok_or_else(|| Error::Internal(format!("Field '{}' not found", field_name)).extend()),
+        _ => Err(Error::Internal("Unexpected Move struct type".to_string()).extend()),
+    }
+}
+
+#[Object]
+impl DynamicField {
+    async fn name(&self, ctx: &Context<'_>) -> Result<Option<MoveValue>> {
```

Comment:
I can see why this was frustrating to you -- I have some ideas for how to clean this up, but I'll tackle them afterwards, so will just focus on correctness in this review.